File: src/features/typing/data/ts/kth_smallest_element_in_a_bst.ts
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function kthSmallest(root: TreeNode | null, k: number): number {
    const stack: TreeNode[] = [];
    let cur: TreeNode | null = root;
    while (cur !== null || stack.length > 0) {
        while (cur !== null) {
            stack.push(cur);
            cur = cur.left;
        }
        cur = stack.pop()!;
        k--;
        if (k === 0) {
            return cur.val;
        }
        cur = cur.right;
    }
    return -1;
}

File: src/features/typing/data/dart/kth_smallest_element_in_a_bst.dart
class Solution {
  int kthSmallest(TreeNode? root, int k) {
    final stack = <TreeNode>[];
    TreeNode? cur = root;
    while (cur != null || stack.isNotEmpty) {
      while (cur != null) {
        stack.add(cur);
        cur = cur.left;
      }
      cur = stack.removeLast();
      k--;
      if (k == 0) {
        return cur.val;
      }
      cur = cur.right;
    }
    return -1;
  }
}
